arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -508,6 +440,168 @@ public static VectorizedReader<?> replaceWithMetadataReader(
+              new LogicalTypeVisitorResult(
+                  vector, ReadType.LONG_BACKED_DECIMAL, (int) BigIntVector.TYPE_WIDTH));
+        case INT32:
+          ((IntVector) vector).allocateNew(batchSize);
+          return Optional.of(
+              new LogicalTypeVisitorResult(
+                  vector, ReadType.INT_BACKED_DECIMAL, (int) IntVector.TYPE_WIDTH));
+        default:
+          throw new UnsupportedOperationException(
+              "Unsupported base type for decimal: " + primitive.getPrimitiveTypeName());
+      }
+    }
+
+    @Override
+    public Optional<LogicalTypeVisitorResult> visit(
+        LogicalTypeAnnotation.DateLogicalTypeAnnotation dateLogicalType) {
+      FieldVector vector = arrowField.createVector(rootAlloc);
+      ((DateDayVector) vector).allocateNew(batchSize);
+      return Optional.of(
+          new LogicalTypeVisitorResult(vector, ReadType.INT, (int) IntVector.TYPE_WIDTH));
+    }
+
+    @Override
+    public Optional<LogicalTypeVisitorResult> visit(
+        LogicalTypeAnnotation.TimeLogicalTypeAnnotation timeLogicalType) {
+      FieldVector vector = arrowField.createVector(rootAlloc);
+      ((TimeMicroVector) vector).allocateNew(batchSize);
+      return Optional.of(
+          new LogicalTypeVisitorResult(vector, ReadType.LONG, (int) TimeMicroVector.TYPE_WIDTH));
+    }
+
+    @Override
+    public Optional<LogicalTypeVisitorResult> visit(
+        LogicalTypeAnnotation.TimestampLogicalTypeAnnotation timestampLogicalType) {
+      FieldVector vector = arrowField.createVector(rootAlloc);
+      switch (timestampLogicalType.getUnit()) {
+        case MILLIS:
+          ((BigIntVector) vector).allocateNew(batchSize);
+          return Optional.of(
+              new LogicalTypeVisitorResult(
+                  vector, ReadType.TIMESTAMP_MILLIS, (int) BigIntVector.TYPE_WIDTH));
+        case MICROS:
+          if (((Types.TimestampType) icebergField.type()).shouldAdjustToUTC()) {
+            ((TimeStampMicroTZVector) vector).allocateNew(batchSize);
+          } else {
+            ((TimeStampMicroVector) vector).allocateNew(batchSize);
+          }
+
+          return Optional.of(
+              new LogicalTypeVisitorResult(vector, ReadType.LONG, (int) BigIntVector.TYPE_WIDTH));
+        case NANOS:
+          if (((Types.TimestampNanoType) icebergField.type()).shouldAdjustToUTC()) {
+            ((TimeStampNanoTZVector) vector).allocateNew(batchSize);
+          } else {
+            ((TimeStampNanoVector) vector).allocateNew(batchSize);
+          }
+
+          return Optional.of(
+              new LogicalTypeVisitorResult(vector, ReadType.LONG, (int) BigIntVector.TYPE_WIDTH));
+      }
+
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<LogicalTypeVisitorResult> visit(
+        LogicalTypeAnnotation.IntLogicalTypeAnnotation intLogicalType) {
+      FieldVector vector = arrowField.createVector(rootAlloc);
+
+      if (intLogicalType.getBitWidth() == 8
```

Comment:
maybe extract `intLogicalType.getBitWidth()` into a var so that the code gets shorter